### DISCUSSION TITLE
Hard-coding button mapping on a per-system basis.

### DISCUSSION THREAD

**Discussion by vicsidious** [darthcloud/BlueRetro#433] [Ideas]
Hi! What an amazing project!! 
         I've read about every discussion topics to date and I've stumbled upon a few people asking if they could use mapping profiles. This would be a feature I'd enjoy very much, as well. Even though this seems to be a very coveted feature, the short answer is "no", for now. 
But sometimes, an issue about mapping is raised and @darthcloud fixes it by correcting mapping in the firmware.

Since the code is open source, would it be possible to change mapping per-system, in the firmware ourselves, then compiling it for a personal use and flashing it to our ESP32? Worst case, the fw compiles but has a bug, and we can reset the ESP32 or flash an official fw back to it! Of course the unofficial fw would be unsupported on your side!

One a more technical note: What classes would we have to update to do so? And finally, how would one compile the source under a Windows environment? 

Thanks a lot for the amazing work! I look forward to your answer! 
-David

**Comment by vicsidious** [comment]
I suspect that particular portion of the code to be under main\adapter\wire(d/less)\<console>.c 
Could moving around values in the <console>_btns_mask array achieve what I want to? 

Ex:
![image](https://user-images.githubusercontent.com/115555660/195724458-cb9414dc-3f3a-45bc-8e8c-8ce4c6a5714a.png)

I would give it a go and test it gladly, but I have absolutely no idea if it can be built in a Windows environment...

**Comment by vicsidious** [comment]
Ok, I've moved some of the bits around in the aforementioned saturn class, just to try, and now the only thing left was the daunting task of building the firmware. Now, I've read the wiki on IDE setup...

Basically, I: 
1) Bought a new laptop 
2) Installed Ubuntu
3) Locally cloned the repo
4) Swapped what I thought was the controls of the saturn (as proposed in the last comment)
5) And after a few summersaults with Python not being recognised, I finally was able to BUILD it! 
Now, can I just flash the resulting BlueRetro/Build/BlueRetro.bin file on my ESP32, or is there something else to do before that? 

Please @darthcloud if you could shed some light on this for me, it would mean a lot! 
Merci beaucoup
-David

**Comment by vicsidious** [comment]
VICTORY!!! 
I did swap the Saturn Left/Right/Up/Down buttons around, and flashed the resulting bin to the ESP32 and it worked! I tried on the Sega Genesis and the DPad remained untouched <3 

Looking at all that code, all the different managed situations, I am amazed at how efficient a dev you are Darthcloud! 
Congratulation on such an amazing project!

**Reply by darthcloud** [reply]
Congrats, glad you figured it out. 😎

**Reply by vicsidious** [reply]
Clean code makes it SO easy to navigate! Congratulations are all yours, mate! :)
